sei-tendermint/internal/p2p/peermanager.go

```diff
@@ -4,1621 +4,527 @@ import (
-// subscribers. The peer must already be marked as connected. This is
-// separate from Dialed() and Accepted() to allow the router to set up
-// its internal queues before reactors start sending messages. The
-// channels set here are passed in the peer update broadcast to
-// reactors, which can then mediate their own behavior based on the
-// capability of the peers.
-func (m *PeerManager) Ready(ctx context.Context, peerID types.NodeID, channels ChannelIDSet) {
-	m.mtx.Lock()
-	defer m.mtx.Unlock()
-
-	if m.connected[peerID] {
-		for ready, ctrl := range m.ready.Lock() {
-			ready[peerID] = channels
-			ctrl.Updated()
-		}
-		m.broadcast(ctx, PeerUpdate{
-			NodeID:   peerID,
-			Status:   PeerStatusUp,
-			Channels: channels,
-		})
+func (i *peerManagerInner[C]) Evict(id types.NodeID) {
+	if !i.isPersistent(id) {
+		delete(i.addrs, id)
 	}
-}
-
-// EvictNext returns the next peer to evict (i.e. disconnect). If no evictable
-// peers are found, the call will block until one becomes available.
-func (m *PeerManager) EvictNext(ctx context.Context) (Eviction, error) {
-	for {
-		ev, err := m.TryEvictNext()
-		if err != nil {
-			return Eviction{}, err
-		}
-		if ev, ok := ev.Get(); ok {
-			return ev, nil
-		}
-		select {
-		case <-m.evictWaker.Sleep():
-		case <-ctx.Done():
-			return Eviction{}, ctx.Err()
+	if !i.isUnconditional[id] {
+		if conn, ok := i.conns.Load().Get(id); ok {
+			conn.Close()
 		}
 	}
 }
 
-type Eviction struct {
-	ID    types.NodeID
-	Cause error
-}
-
-// TryEvictNext is equivalent to EvictNext, but immediately returns an empty
-// node ID if no evictable peers are found.
-func (m *PeerManager) TryEvictNext() (utils.Option[Eviction], error) {
-	m.mtx.Lock()
-	defer m.mtx.Unlock()
-
-	// If any connected peers are explicitly scheduled for eviction, we return a
-	// random one.
-	for peerID, cause := range m.evict {
-		delete(m.evict, peerID)
-		if m.connected[peerID] && !m.evicting[peerID] {
-			m.evicting[peerID] = true
-			return utils.Some(Eviction{peerID, cause}), nil
+// Connected registers a new connection.
+// If it is an outbound connection the dialing status is cleared (EVEN IF IT RETURNS AN ERROR).
+// It atomically checks maximum connection limit and duplicate connections.
+func (i *peerManagerInner[C]) Connected(conn C) (err error) {
+	defer func() {
+		if err != nil {
+			conn.Close()
 		}
+	}()
+	info := conn.Info()
+	if addr, ok := info.DialAddr.Get(); ok && i.dialing[addr.NodeID] == addr {
+		delete(i.dialing, addr.NodeID)
+	}
+	peerID := info.ID
+	if peerID == i.selfID {
+		return errors.New("connection to self")
+	}
+	conns := i.conns.Load()
+	if old, ok := conns.Get(peerID); ok {
+		// * allow to override connections in the same direction.
+		// * allow inbound conn to override outbound iff peerID > selfID.
+		//   This resolves the situation when peers try to connect to each other
+		//   at the same time.
+		oldDir := old.Info().DialAddr.IsPresent()
```

Comment:
shouldn't this be `oldAddr` or something instead of `oldDir`? (same for `newDir`)